slog context support

pohly opened the pull request:
Conversion is done on demand. A program which uses either slog or logr for
logging remains as efficient as before. What is less efficient is storing a
logger of one type and then retrieving it as the other.
    
Ideally, reusable packages should avoid using a slog.Logger and prefer
logr.Logger instead when retrieving a logger from a context. There is a
significant amount of packages which already work that way (Kubernetes and
related packages), while storing and retrieving a slog.Logger is new and not
done anywhere yet. Emulating existing behavior ensures better performance.

To enable on demand conversion, the slogr implementation must be moved
into the main package.

Related-to: https://github.com/go-logr/logr/issues/234

pohly commented on context_slog.go:
```diff
@@ -0,0 +1,83 @@
+}
+
+// FromContext returns a slog.Logger from ctx or nil if no such Logger is found.
+func SlogLoggerFromContext(ctx context.Context) *slog.Logger {
```
I decided to focus on handling a `slog.Logger` pointer instead of a `slog.Handler` because a) that is what [slog-context](https://github.com/veqryn/slog-context) does and b) because only supporting `slog.Handler` would imply one `slog.New` on each retrieval.

It would also be possible to extend this to `slog.Handler`.

pohly replied:
In a way, the whole point of context support for slog is to enable a familiar API for developers - that is `slog.Logger`, not `slog.Handler`.

I still think developers should be more flexible and just use `logr.Logger`... :shrug:

veqryn replied:
👍

thockin reviewed (commented):
This isn't horrible, to me, but it seems like one day slog will decide to add context support and then we'll have to do it all again.

thockin commented on context_noslog.go:
```diff
@@ -0,0 +1,49 @@
+//go:build !go1.21
```
sigh

thockin commented on slogr.go:
```diff
@@ -0,0 +1,100 @@
+//	slog.New(NewSlogHandler(logger)).Info(...) -> logger.GetSink().Info(level=0, ...)
+//	slog.New(NewSlogHandler(logger.V(4))).Info(...) -> logger.GetSink().Info(level=4, ...)
+func NewSlogHandler(logger Logger) slog.Handler {
+	if sink, ok := logger.GetSink().(*slogSink); ok && logger.GetV() == 0 {
```
I was just satying that `SlogSink` is the name of the type, but maybe I made a mistake

pohly replied:
The name is right as it is: `slogSink` is the internal helper struct.

pohly commented:
I've pushed an update:
- first commit updated to restore `slogr/slogr_test.go`
- new commits for renaming and documentation

pohly commented on README.md:
```diff
@@ -145,11 +146,12 @@ There are implementations for the following logging libraries:
 Interoperability goes both ways, using the `logr.Logger` API with a `slog.Handler`
-and using the `slog.Logger` API with a `logr.LogSink`. [slogr](./slogr) provides `NewLogr` and
-`NewSlogHandler` API calls to convert between a `logr.Logger` and a `slog.Handler`.
+and using the `slog.Logger` API with a `logr.LogSink`. [slogr](./slogr) provides `NewLogr` (= `logr.NewFromSlogHandler`) and
```
I wasn't sure whether we'd want to deprecate slogr - it just got added :hot_face: 

But we can also do that, no strong opinion from me about it.

thockin replied:
I don't love having two equivaent ways of saying the same thng which are both "suported".  Deprecated doesn't mean gone, just "look here instead".  IMO

pohly replied:
Makes sense.

pohly commented on slogr/slogr.go:
```diff
@@ -36,14 +36,11 @@ import (
+//
+// NewLogr is identical to [logr.NewFromSlogHandler]. Either of these can
+// be used.
 func NewLogr(handler slog.Handler) logr.Logger {
```
I don't like that because conceptually, a `var NewLogr` would allow users to replace the function. Not that they'll ever do it, but the API allows it, which feels wrong.

Regarding performance: I'm not sure whether it matters and whether it really makes a difference. It could also backfire because the Go compiler no longer inline the code because `var NewLogr` could be anything.

I'd love to have a `const NewLogr = logr.NewFromSlogHandler`, but there is no such thing in Go.

pohly replied:
I got curious and benchmarked both approaches. No significant difference.

thockin replied:
Yeah, I didn't mean benchmark, just one less call frame to debug thru.  You're right that mutability is unfortunate here.  OK.

pohly commented on slogr/slogr.go:
```diff
@@ -36,14 +36,11 @@ import (
 // slog.LevelInfo and V(4) becomes slog.LevelDebug.
+//
+// NewLogr is identical to [logr.NewFromSlogHandler]. Either of these can
+// be used.
```
The duplicated docs are indeed annoying. I'm leaning towards simply deprecating `slogr` now.

thockin replied:
+1

pohly commented on context_slog.go:
```diff
@@ -0,0 +1,83 @@
+}
+
+// FromContext returns a slog.Logger from ctx or nil if no such Logger is found.
+func SlogLoggerFromContext(ctx context.Context) *slog.Logger {
```
Yes, make sense.

pohly commented on slogr_test.go:
```diff
@@ -0,0 +1,261 @@
+	// level=DEBUG msg="with values, verbosity and name" x=1 y=2 str=abc logger=foo/bar
+}
+
+func ExampleNewSlogLogger() {
```
Yes, fixed.

pohly commented on slogr_test.go:
```diff
@@ -0,0 +1,261 @@
+}
+
+func ExampleNew() {
+	logger := logr.NewFromSlogHandler(slog.NewTextHandler(os.Stdout, debugWithoutTime))
```
Done.

thockin commented on slogr_test.go:
```diff
@@ -0,0 +1,261 @@
+limitations under the License.
+*/
+
+package logr_test
```
even if it is `func X() { return log.X() }`, the test seems redundant.

pohly replied:
I'll remove the test.

But it did find a problem earlier: instead of the `type SlogSink = logr.SlogSink`, I had a copy because I wanted it to be visible in the godoc. That normally works when methods are the same, but here the type is part of the interface, and that made the two interfaces different. Anyway, that's not a problem anymore.

thockin commented:
ICYMI - there was no new push :)

pohly commented:
> ICYMI - there was no new push :)

I wanted to revise the documentation part first regarding deprecation, but I am not going to get to that this evening anymore, so I pushed what I have just now. The next update will be tomorrow.

pohly commented:
I've updated the documentation and removed the `slogr/slogr_test.go`: https://github.com/go-logr/logr/compare/9ac998326cdc223355affa6baeab839316117be4..ca4a15b75ca78b8f2db2de3e41b660c810ef5b9e

pohly commented on README.md:
```diff
@@ -208,24 +209,36 @@ with logr implementations without slog support is not important, then
+provides additional support code for this approach. It also contains wrappers
+for the context functions in logr, so developers who prefer to not use the logr
+APIs directly can use those instead and the resulting code will still be
+interoperable with logr.
```
@veqryn: does this look okay to you?

veqryn replied:
Looks good to me. Thank you.

thockin commented on README.md:
```diff
@@ -145,15 +146,15 @@ There are implementations for the following logging libraries:
 Interoperability goes both ways, using the `logr.Logger` API with a `slog.Handler`
-and using the `slog.Logger` API with a `logr.LogSink`. [slogr](./slogr) provides `NewLogr` and
-`NewSlogHandler` API calls to convert between a `logr.Logger` and a `slog.Handler`.
+and using the `slog.Logger` API with a `logr.LogSink`. `NewFromSlogHandler` and
```
ISTR an earlier draft of this used names like `ToSlog` and `FromSlog` or `{To,From}SlogHandler` - now that I have come around, do you want to reconsider those?

thockin commented on README.md:
```diff
@@ -145,15 +146,15 @@ There are implementations for the following logging libraries:
+slog API.
 
-## Using a `logr.Sink` as backend for slog
+### Using a `logr.Sink` as backend for slog
```
the type is `logr.LogSink`

thockin reviewed (commented):
LGTM overall

thockin commented:
LGTM - hold for @veqryn to take a look?

veqryn reviewed (approved):
👍  LGTM
After the next release, I will update slog-context to use this, and pull you all into the pull PR to take a look. :)

veqryn commented:
@pohly @thockin Could I get your eyes on this pr please: https://github.com/veqryn/slog-context/pull/1

pohly commented on slogr/slogr.go:
```diff
@@ -23,86 +23,32 @@ limitations under the License.
-	}
-	return handler
+// Deprecated: use [logr.ToSlogHandler] instead.
+func ToSlogHandler(logger logr.Logger) slog.Handler {
```
Argh!

I messed up some global search/replace and renamed `NewSlogHandler` to `ToSlogHandler` although here we have to keep the old name.

Found when updating logr in klog. Will prepare a fix.
